File: src/games/buttons.py
"""Buttons activity - press every themed button once.

Port of the spirit of the legacy ``pysybuttons.py``: a set of themed objects is
laid out on screen, each with a "resting" and a "pressed" picture (``NN.png`` /
``NNb.png`` in the theme folder).  Clicking an object presses it; press them all
to win the round.
"""

from __future__ import annotations

import random

import pygame

from .base import (
    LEVEL_EASY,
    LEVEL_HARD,
    LEVEL_MEDIUM,
    MARGIN_LEFT,
    MARGIN_TOP,
    PLAY_HEIGHT,
    PLAY_WIDTH,
    Activity,
)

BUTTON_COUNT = {LEVEL_EASY: 3, LEVEL_MEDIUM: 5, LEVEL_HARD: 8}
BUTTON_MAX_EDGE = 120


class _Button:
    __slots__ = ("rest", "pressed", "rect", "is_pressed")

    def __init__(self, rest, pressed, center) -> None:
        self.rest = rest
        self.pressed = pressed
        self.rect = rest.get_rect(center=center)
        self.is_pressed = False

    @property
    def image(self):
        return self.pressed if self.is_pressed else self.rest


class ButtonsActivity(Activity):
    id = "buttons"
    title = "Buttons"
    background = "buttons"
    themes_suffix = "buttons"

    def __init__(self, engine, level=LEVEL_MEDIUM) -> None:
        super().__init__(engine, level)
        self.buttons: list[_Button] = []

    # ------------------------------------------------------------------
    def _button_pairs(self) -> list[tuple]:
        """Return (rest_path, pressed_path_or_None) pairs from the theme folder."""
        by_stem = {p.stem: p for p in self.theme_images(exclude_prefixes=("fond", "logo"))}
        pairs = []
        for stem, path in sorted(by_stem.items()):
            if stem.endswith("b"):
                continue
            pairs.append((path, by_stem.get(stem + "b")))
        return pairs

    def _scaled(self, path) -> pygame.Surface:
        img = self.load_theme_image(path)
        w, h = img.get_size()
        scale = min(1.0, BUTTON_MAX_EDGE / max(w, h))
        if scale < 1.0:
            img = pygame.transform.smoothscale(img, (int(w * scale), int(h * scale)))
        return img

    def _fallback(self, pressed: bool) -> pygame.Surface:
        surf = pygame.Surface((90, 60), pygame.SRCALPHA)
        colour = (90, 170, 90) if pressed else (170, 170, 90)
        pygame.draw.rect(surf, colour, surf.get_rect(), border_radius=10)
        return surf

    def start_round(self) -> None:
        pairs = self._button_pairs()
        random.shuffle(pairs)
        want = BUTTON_COUNT.get(self.level, 5)
        cols = min(want, 4)
        rows = (want + cols - 1) // cols
        cell_w = PLAY_WIDTH // cols
        cell_h = PLAY_HEIGHT // (rows + 1)

        self.buttons = []
        for i in range(want):
            if pairs:
                rest_path, pressed_path = pairs[i % len(pairs)]
                try:
                    rest = self._scaled(rest_path)
                    pressed = self._scaled(pressed_path) if pressed_path else rest
                except pygame.error:
                    rest, pressed = self._fallback(False), self._fallback(True)
            else:
                rest, pressed = self._fallback(False), self._fallback(True)
            col, row = i % cols, i // cols
            center = (
                MARGIN_LEFT + cell_w * col + cell_w // 2,
                MARGIN_TOP + cell_h * (row + 1) + cell_h // 2,
            )
            self.buttons.append(_Button(rest, pressed, center))

    on_theme_changed = start_round

    # ------------------------------------------------------------------
    def handle_event(self, event: pygame.event.Event) -> None:
        if event.type != pygame.MOUSEBUTTONDOWN or event.button != 1:
            return
        for button in reversed(self.buttons):
            if not button.is_pressed and button.rect.collidepoint(event.pos):
                button.is_pressed = True
                self.score += 1
                self.engine.play_sound("btncoche.ogg")
                break

    def is_round_won(self) -> bool:
        return bool(self.buttons) and all(b.is_pressed for b in self.buttons)

    def draw(self, surface: pygame.Surface) -> None:
        surface.blit(self.bg_surface, (0, 0))
        for button in self.buttons:
            surface.blit(button.image, button.rect)

File: web/js/games/buttons.js
// Buttons activity -- press every themed button once.
// Counterpart of src/games/buttons.py.

import { C } from "../engine.js";
import { Activity, MARGIN_LEFT, MARGIN_TOP, PLAY_H, PLAY_W, fitDown, shuffle } from "./base.js";

export class ButtonsActivity extends Activity {
  static id = "buttons";
  static suffix = "buttons";
  static bg = "buttons";
  static title = "Buttons";

  themeAssetList(data) {
    const out = [data.logo];
    for (const b of data.buttons || []) out.push(b.rest, b.pressed);
    return out.filter(Boolean);
  }

  startRound() {
    const pairs = shuffle([...(this.themeData.buttons || [])]);
    const want = C.BUTTON_COUNT[this.level] ?? 5;
    const cols = Math.min(want, 4);
    const rows = Math.ceil(want / cols);
    const cellW = PLAY_W / cols;
    const cellH = PLAY_H / (rows + 1);

    this.buttons = [];
    for (let i = 0; i < want; i++) {
      const pair = pairs.length ? pairs[i % pairs.length] : null;
      const rest = pair && this.engine.image(pair.rest);
      const pressed = (pair && this.engine.image(pair.pressed)) || rest;
      const nw = rest?.naturalWidth || 90;
      const nh = rest?.naturalHeight || 60;
      const { w, h } = fitDown(nw, nh, C.BUTTON_MAX_EDGE);
      const col = i % cols;
      const row = Math.floor(i / cols);
      const cx = MARGIN_LEFT + cellW * col + cellW / 2;
      const cy = MARGIN_TOP + cellH * (row + 1) + cellH / 2;
      this.buttons.push({ rest, pressed, x: cx - w / 2, y: cy - h / 2, w, h, on: false });
    }
  }

  onPointer(ev) {
    if (ev.type !== "pointerdown" || ev.button !== 0) return;
    for (let i = this.buttons.length - 1; i >= 0; i--) {
      const b = this.buttons[i];
      if (!b.on && ev.x >= b.x && ev.x <= b.x + b.w && ev.y >= b.y && ev.y <= b.y + b.h) {
        b.on = true;
        this.score++;
        this.engine.sound?.("sounds/btncoche.ogg");
        break;
      }
    }
  }

  isRoundWon() {
    return this.buttons && this.buttons.length > 0 && this.buttons.every((b) => b.on);
  }

  paint(ctx) {
    for (const b of this.buttons) {
      const img = b.on ? b.pressed : b.rest;
      if (img) ctx.drawImage(img, b.x, b.y, b.w, b.h);
      else {
        ctx.fillStyle = b.on ? "#5aaa5a" : "#aaaa5a";
        ctx.fillRect(b.x, b.y, b.w, b.h);
      }
    }
  }
}
